### ISSUE TITLE
Add available memory state for hostmetrics receiver

### ISSUE THREAD

**Issue by rogercoll** [open-telemetry/opentelemetry-collector-contrib#7417]
The current Linux memory states provided by the `hostmetricsreceiver` might be not the most appropriate ones for an end user. The available memory states for Linux are:

`[buffered, cached, free, slab_reclaimable, slab_unreclaimable, used] `

All of those values are retrieved with gopsutil, which gets them from `/proc/meminfo` file.

A user might think that the `free` value represents the amount of memory that the system can use, but free (MemFree) is a count of how many pages are free in the buddy allocator, however there are a lot of ways pages can be returned to the buddy allocator in time of need. The latest is provided by the Linux kernel (>3.16) as `MemAvailable` in `/proc/meminfo`.

**Can we get MemAvailable field from current states?**

No, it is an independent field not computed with the current values. Gopsutils provides this value regardless the kernel version as it can be [estimated](https://unix.stackexchange.com/questions/261247/how-can-i-get-the-amount-of-available-memory-portably-across-distributions/261252#261252) from MemFree, Active(file), Inactive(file), and SReclaimable, as well as the "low" watermarks from /proc/zoneinfo.

**Repercussion**

The user will be able to know the actual available memory of the system.

- Will it break the assumption that the sum of all metrics states should be equal to the limit (total memory)? https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-naming

It is already broken and quite difficult to achieve it as some of the memory state values are inclusive, for example the [cached state](https://github.com/shirou/gopsutil/blob/master/mem/mem_linux.go#L304). The assumption would be defined as:

`total = buffered + cached + free + used + slab_reclaimable + slab_unreclaimable`

But the `used` value is computed by gopsutils as: [total - free - buffered - cached](https://github.com/shirou/gopsutil/blob/master/mem/mem_linux.go#L314)

Then:

`total = buffered + cached + free + (total - free - buffered - cached) + slab_reclaimable + slab_unreclaimable`

`total = total + slab_reclaimable + slab_unreclaimable`

The sum(state) will always be higher than the total.

(The following proposals do not solve the sum assumption, they only add more memory information for the user)

**Proposal 1**

Add a new memory state for Linux to provide the available memory. State name: `avaialble`

**Proposal 2**

Change `used` state for: `total - available`
This proposal requires to add the memory.limit metric as we won't be able to compute the total memory from the states. Currently, we can get the total amount of memory with the following promQL query: `scalar(sum without (state) (system_memory_usage{state =~ "free|used|cached|buffered"}))`

**Comment by jpkrohling** [comment]
cc @dmitryax

**Comment by quentinmit** [comment]
I don't think either of those proposals improves the problem with the values not adding up. I'd like to add proposal 3, fixing the set of states to exactly cover the memory on the system. If there are states that double-count memory, they should be separate metrics. Here are the non-overlapping states from Munin and how they [calculate](https://github.com/munin-monitoring/munin/blob/master/plugins/node.d.linux/memory#L292) them from /proc/meminfo as an inspiration:
```
apps (MemTotal-MemFree-Buffers-Cached-Slab-PageTables-Percpu-SwapCached)
buffers (Buffers)
swap (SwapTotal-SwapFree)
cached (Cached-Shmem)
free (MemFree)
shmem (Shmem)
slab_cache (Slab)
swap_cache (SwapCached)
page_tables (PageTables)
per_cpu (Percpu)
vmalloc_used (VmallocUsed)
```

(I'm not sure if it makes sense to include swap; I'm guessing that with the current set of states we're not including swap? You could leave that out and I think the remaining states would represent total physical memory.)

I think the main difference between this and proposal 2 is that here we're changing the definition of `used` (what Munin calls `apps`) instead of changing the definition of `free`.

**Comment by rogercoll** [comment]
Another possible solution would be to add a new memory metric only for this purpose, as making any change on the current states would break the sum assumption. 

The memory system scrapper will generate the following metrics:

```
- system.memory.usage: state attributes
- system.memory.utilization: disabled by default, state attributes
- system.memory.available: no state attributes
```

Any thoughts?

**Comment by TylerHelmuth** [comment]
> Another possible solution would be to add a new memory metric only for this purpose 

Is this the same as your proposed solution 1?

I like the idea of adding a new metric; allows those who currently rely on `used` to continue to use it, while providing a more accurate measure for those who want it.

**Comment by rogercoll** [comment]
@TylerHelmuth No, thanks for pointing that out. It would be a new proposal, let me update the comment. 

Proposals 1 and 2 are based on the "state" attribute used in usage and utilization metrics. Proposal 3 won't do any modification on the state neither the current metrics, instead, add a new one.

**Comment by dmitryax** [comment]
Adding another metric `system.memory.available` maybe be a good solution to one the problems described in this issue. But it's still unclear how to solve the total mismatch issue and also adding this metric brings confusion to end user questioning what's the difference between `system.memory.available` and `system.memory.usage`.

Potentially, we can solve both problems by introducing a concept of optional metric attribute values to the metrics builder. The attributes values can be enabled or disabled in metadata.yaml (default) and user configs. If attribute value is disabled, it means that datapoints with this attribute value won't be emitted.

With this functionality added to the metrics builder, we will have the following sets of `state` attribute values in this receivers:
- Enabled by default: `buffered`, `cached`, `free` and `used`. Sums up to a correct total value.
- Disabled by default: `slab_reclaimable`, `slab_unreclaimable` and `available`. Optional values that user can enable.

What do you think?

@bogdandrutu @djaglowski please share your opinion on the additional feature to the metrics builder

**Comment by djaglowski** [comment]
I think we should aim for a design that adheres to the specification's aggregation expectations (ie. data points sum to a meaningful total). Supporting metrics that violate these expectations seems like the wrong approach to me. 

There seem to be two layers to the problem here:
1. What is the correct data model for this situation?
2. How do we migrate from a flawed data model to the correct one?

I don't have a strong opinion on the first question, but in regards to the second, I think we basically have to decide whether to fix it, or to deprecate and replace it. The right approach depends on stability level. In my opinion, a per-metric stability level is necessary in order to have a clear approach (eg. alpha -> beta -> stable -> deprecated). If this metric is less than stable, then we should fix it. If it is "stable" but realized to be fundamentally flawed in its design, we should deprecate it and add one or more new metrics that amount to a more accurate model.

**Comment by dmitryax** [comment]
> I think we should aim for a design that adheres to the specification's aggregation expectations (ie. data points sum to a meaningful total). Supporting metrics that violate these expectations seems like the wrong approach to me.

That's true, we should always aim to have meaningful total values. Proposing optional attribute values, I imply that the sum of attribute values enabled by default should represent the most reasonable total value. At the same time, I can think of use cases when user wants to disable some attributes values or add others optional ones and have their own custom meaningful sums out of them. In this case moving `slab_reclaimable` and `slab_unreclaimable` to a different metric would significantly reduce flexibility because they are part of the same memory usage metric.

**Comment by quentinmit** [comment]
I think it's okay to keep `SlabReclaimable` and `SlabUnreclaimable` in the regular summable metric; their sum is just `Slab`. So you can include either `slab` or `slab_reclaimable` and `slab_unreclaimable` but not both. Other states are the problematic ones.

**Comment by lootek** [comment]
I think this issue is still valid and worth addressing (@dmitryax) as we're clearly missing `system.memory.available` metrics.

And so I'd like to reopen the discussion as to me it seems like adding `system.memory.available` is what all of us agree to do (especially if it's just a matter of using a value already exposed by `gopsutil`).

As for fixing the flawed sum, I like this approach:
> With this functionality added to the metrics builder, we will have the following sets of `state` attribute values in this receivers:
> 
> * Enabled by default: `buffered`, `cached`, `free` and `used`. Sums up to a correct total value.
> * Disabled by default: `slab_reclaimable`, `slab_unreclaimable` and `available`. Optional values that user can enable.

**Comment by lootek** [comment]
@fatsheep9146 maybe I could take this one and try implementing those changes?

**Comment by fatsheep9146** [comment]
@lootek It's yours!

**Comment by lootek** [comment]
I have a [draft PR prepared](https://github.com/lootek/opentelemetry-collector-contrib/pull/1) with the `available` memory state attribute added, yet this is for now the implementation of the simplest solution that just adds this state. 

As for making some of the attributes optional, since I'm new to the codebase and don't yet have a full grasp of it I'd love to hear feedback from you and ask you for some initial guidance.
I spent quite a long time staring at the `metadata.yaml` file and a few solutions came to my mind yet none of those I'm happy with... Particularly because I was focused on a solution that would be fully backward-compatible but all of those required some changes to `mdatagen` by at least adding another field aside `enum`.
The one I can see that could require changes only to the memoryscraper itself was Inspired by filters available in eg. `filesystemscraper` configuration. Eventually, I landed in the file containing [`AttributesMatcher`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95625e0303bdfa7b51e0896fdce301fe336548f0/internal/filter/filtermatcher/attributematcher.go) implementation - do you think it would be any useful here? 

If the changes to `mdatagen` turn out to be inevitable, to my understanding the desired order of work would be to first do those in a separate PR and then issue another PR addressing the particular case of available memory using those changes?

**Comment by lootek** [comment]
It looks like some related work has been merged just a few hours ago:
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17967

Given that I wonder if we still need a special logic for grouping/choosing memory `state` attributes here. In case we do, I've prepared a PR (built on top of that previous one) that adds a config option for that - please take a look:
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/19149

**Comment by lootek** [comment]
The discussion follows in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/19149

**Comment by ItsLastDay** [comment]
I propose to walk the "short-term" path. I created https://github.com/open-telemetry/opentelemetry-specification/issues/3659.

**Comment by ItsLastDay** [comment]
https://github.com/open-telemetry/semantic-conventions/issues/257 is done. Next step: I'll implement `system.linux.memory.available` in `hostmetricsreceiver`.

**Comment by ItsLastDay** [comment]
I implemented all steps for "Issue B" from https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7417#issuecomment-1668082340.

So, answering the original question from this bug - "Can we get MemAvailable field from current states?": now yes. Thus, I'll close this bug (I hope I can). 

Feel free to open a separate thread for "Issue A" from https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7417#issuecomment-1668082340. I see that https://github.com/open-telemetry/semantic-conventions/issues/531 is still open, and IMO it's not critical.

**Comment by quentinmit** [comment]
> Feel free to open a separate thread for "Issue A" from [#7417 (comment)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7417#issuecomment-1668082340). I see that [open-telemetry/semantic-conventions#531](https://github.com/open-telemetry/semantic-conventions/issues/531) is still open, and IMO it's not critical.

The linked semconv issue doesn't address this bug - it doesn't fix the `free` state to represent something meaningful, and it doesn't propose fixing anything about the existing metric. I think we should keep using this issue for that, given that most of the comments are about it.

**Comment by rogercoll** [comment]
I would keep this issue open too, the existing memory metrics won't be aligned with the semantic conventions until the `slab` state is replaced.

SemConv progress update: https://github.com/open-telemetry/semantic-conventions/pull/1078

**Comment by rogercoll** [comment]
/unstale `system.linux.memory.slab` metric will be available in the next semconv release: https://github.com/open-telemetry/semantic-conventions/pull/1078

**Comment by rogercoll** [comment]
@open-telemetry/collector-contrib-triagers "never-stale"?

**Comment by crobert-1** [comment]
@rogercoll Can you provide an update on the status of this issue? (It looks like the semantic convention PRs have been merged, I'm not sure what's left/required for this issue to be complete at this point)

**Comment by rogercoll** [comment]
> [@rogercoll](https://github.com/rogercoll) Can you provide an update on the status of this issue? (It looks like the semantic convention PRs have been merged, I'm not sure what's left/required for this issue to be complete at this point)

Default memory state attributes have been updated in SemConv: https://github.com/open-telemetry/semantic-conventions/pull/2534/files. `used` state now relies on Linux `MemAvailable` value by default: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/42223

The sum of the current memory state attributes is still incorrect, it is not equal to the `limit`. Nonetheless, I am closing this issue as this will be fixed with System SemConv stabilization in the `hostmetrics` receiver.

To sum up, `available` memory state is now the default in `system.memory.usage` metric.
